### PR TITLE
Refactored iterator module

### LINKED ISSUE #31: Replace `unwrap()`
There are many places where `.unwrap()` is currently called.
These should be removed or replaces with the correct error handing or `.expect(...)` if it is very clear that it could not be triggered.

This is not needed for tests, `unwrap` is fine in those cases.
The effort should be mainly focused on the `rrule/src/iter` and `rrule/src/parser` folder.

### LINKED ISSUE #61: `RRuleSetIter` doesn't elaborate `RRuleIter` errors
This code panics:

```
#[test]
fn edge_case() {
    use rrule::DateFilter;

    let rrule: RRuleSet = "DTSTART;TZID=Europe/Berlin:18930401T010000;\nRRULE:FREQ=DAILY"
        .parse()
        .expect("The RRule is not valid");

    dbg!(rrule.all_with_error(10));
}
```

I need to change [these lines](https://github.com/fmeringdal/rust-rrule/blob/main/rrule/src/iter/rrule_iter.rs#L168) to:
```
let res = date
                .and_hms_opt(0, 0, 0)
                .ok_or_else(|| RRuleError::new_iter_err("Invalid datetime."))?
                .checked_add_signed(timeset.duration_from_midnight())
                .ok_or_else(|| RRuleError::new_iter_err("Invalid datetime."))?;
```

But anyway, I cannot see list of errors when I run `cargo test edge_case_2 -- --nocapture`.

PS: If you also panicked and expected the code above should work in the first place and shouldn't panic... check [this](https://github.com/chronotope/chrono/issues/696) out :)

### LINKED ISSUE #62: Fix clippy warnings
Make sure clippy is happy and don't emit any warnings. Then we can also add `cargo clippy -- -D warnings` to the CI.

### LINKED ISSUE #63: Reduce number of type casts
There is a tremendous amount of type casts in this library between different number types. We should try to make these number types more consistent across types so that the code is cleaner and we don't have potential implicit panics.

### PR DESCRIPTION
**Pull request by fmeringdal**
Fixes #31 
Fixes #61 
Fixes #62 
Fixes #63

### CODE REVIEW

**Comment by fmeringdal** [comment]
@omid Another PR if you want to have a look. It is really big, but it solves a couple of issues, among them is the `with_error` issue you reported. Always appreciate your feedback, so wanted to give you a chance to review if you have time :)
Only breaking change this time is the MSRV is bumped to `v1.57.0`

File: rrule/src/parser/error.rs [hunk 1]
```diff
@@ -1,3 +1,4 @@
+#![allow(clippy::module_name_repetitions)]
 use thiserror::Error;
 
 #[derive(Error, Debug, Clone, PartialEq)]
```
Comment (omid) [inline rrule/src/parser/error.rs:4]: I cannot comment down there, and it's not your changes, but around line 64, there is a typo, parameter

File: rrule/src/iter/yearinfo.rs [hunk 2]
```diff
@@ -1,202 +1,204 @@
...
+            .expect("num_days_from_monday is between 0 and 6 which is covered by u16");
+        let mut no1_week_start = pymod(7 - year_start_weekday + rrule_week_start, 7);
+        let first_week_start = no1_week_start;
+        let year_len_ext = if no1_week_start >= 4 {
+            no1_week_start = 0;
+            // Number of days in the year, plus the days we got
+            // from last year.
+            let diff = i32::from(year_start_weekday) - i32::from(rrule_week_start);
+            result.year_len
+                + u16::try_from(pymod(diff, 7)).expect("to be positive because 7 is the modulos")
```
Comment (omid) [inline rrule/src/iter/yearinfo.rs:97]: typo, modulus

File: rrule-afl-fuzz/src/main.rs [hunk 3]
```diff
@@ -6,10 +6,37 @@ use take_rrule::take_rrule_from_data;
...
 use rrule::RRuleSet;
+use std::fs;
+
+// Used for reading the fuzz output
+#[allow(dead_code)]
+fn main2() {
+    let crash_dir = fs::read_dir("out/default/crashes").unwrap();
+    println!("Crashes:");
+    for path in crash_dir {
+        let bytes = std::fs::read(path.unwrap().path()).unwrap();
```
Comment (omid) [inline rrule-afl-fuzz/src/main.rs:17]: `std::` is not required (+ all other places in this file)

File: rrule/src/validator/validate_rrule.rs [hunk 4]
```diff
@@ -27,63 +23,55 @@ pub(crate) static YEAR_RANGE: RangeInclusive<i32> = -10_000..=10_000;
...
-    if let Some(until) = &rrule.until {
-        // Check if before `dt_start`
-        if until < dt_start {
-            return Err(ValidationError::UntilBeforeStart {
-                until: until.to_rfc3339(),
-                dt_start: dt_start.to_rfc3339(),
-            });
-        }
+// Until:
+// - Value should be later then `dt_start`.
```
Comment (omid) [inline rrule/src/validator/validate_rrule.rs:58]: typo, than

**Review by omid — CHANGES_REQUESTED** [review]

**Comment by omid** [comment]
Thanks @fmeringdal.
I couldn't follow all the logic changes. As far as I could follow and checked, couldn't find any issue, just seen perfection!

**Comment by omid** [comment]
And we can also update serde_with to `2.0.0`

File: rrule/src/iter/rrule_iter.rs [hunk 5]
```diff
@@ -58,34 +88,30 @@ impl<'a> RRuleIter<'a> {
         if let Some(err) = self.get_err() {
```
Comment (omid) [inline rrule/src/iter/rrule_iter.rs:88]: Then, I think we can remove the clippy attribute and TODO

File: rrule/src/core/datetime.rs [hunk 6]
```diff
@@ -1,45 +1,32 @@
...
-#[derive(Debug, Clone, Copy)]
-pub(crate) struct Time {
-    pub hour: u8,
-    pub minute: u8,
-    pub second: u8,
-    pub millisecond: u16,
+pub(crate) fn duration_from_midnight(time: NaiveTime) -> Duration {
+    Duration::hours(i64::from(time.hour()))
+        + Duration::minutes(i64::from(time.minute()))
+        + Duration::seconds(i64::from(time.second()))
```
[the remark this answers is not in the document]
Comment (fmeringdal) [reply rrule/src/core/datetime.rs:9]: Yeah, creating `Duration` multiple times isn't ideal, but also not sure what other approach makes it easier. 
Regarding milliseconds, the lowest unit this library deals with is seconds, so adding milliseconds here won't have any effect.

File: rrule/src/iter/rrule_iter.rs (hunk 5 above)
Comment (fmeringdal) [reply rrule/src/iter/rrule_iter.rs:88]: Agreed

**Comment by fmeringdal** [comment]
> Thanks @fmeringdal. I couldn't follow all the logic changes. As far as I could follow and checked, couldn't find any issue, just seen perfection!

The iterator module logic is quite complex, I think mostly because the problem inherently is quite complex and not so much bad code. Thanks for the review! 🙏

**Pull request merged**